### PR TITLE
feat(track) track call improvements for LP2.

### PR DESCRIPTION
**Pull request by Enuviel**

### CODE REVIEW

File: AndroidSDK/src/com/leanplum/ActionContext.java [hunk 1]
```diff
@@ -447,10 +470,52 @@ private String eventWithParentEventNames(String eventName) {
...
+  }
 
-    return fullEventName.toString();
+  /**
+   * Reverse string word by word.
+   *
+   * @param event The string that needs to be reversed.
+   * @return Reversed event string.
+   */
+  private String reverseWordByWord(String event) {
```
Comment (milos1290) [inline AndroidSDK/src/com/leanplum/ActionContext.java]: move this to utils class.

File: AndroidSDK/src/com/leanplum/ActionContext.java [hunk 2]
```diff
@@ -410,35 +426,42 @@ private boolean isChainToExistingMessageStarted(Map<String, Object> args, String
...
       parents.add(context);
       context = context.parentContext;
     }
-    for (int i = parents.size() - 1; i >= -1; i--) {
-      if (fullEventName.length() > 0) {
-        fullEventName.append(' ');
-      }
+
+    for (int i = -1; i < parents.size(); i++) {
+      String messageIgForTrack = messageId;
```
Comment (milos1290) [inline AndroidSDK/src/com/leanplum/ActionContext.java]: `messageIdForTrack`

File: AndroidSDK/src/com/leanplum/ActionContext.java [hunk 3]
```diff
@@ -410,47 +426,90 @@ private boolean isChainToExistingMessageStarted(Map<String, Object> args, String
...
+        context = parents.get(i);
+        actionName = context.key;
+        if (context.chainedToExisting) {
+          messageIdForTrack = context.parentContext.messageId;
+        }
       } else {
         actionName = eventName;
       }
-      if (actionName == null) {
+      if (actionName == null || TextUtils.isEmpty(actionName)) {
```
Comment (milos1290) [inline AndroidSDK/src/com/leanplum/ActionContext.java]: `TextUtils.isEmpty(actionName)` checks for null before checking for length, so `actionName == null` is redundant

File: AndroidSDK/src/com/leanplum/ActionContext.java [hunk 4]
```diff
@@ -410,47 +426,90 @@ private boolean isChainToExistingMessageStarted(Map<String, Object> args, String
...
+        for (Map.Entry<String, StringBuilder> entry : mapOfMessages.entrySet()) {
+          mapOfMessages.put(entry.getKey(), entry.getValue().append(" ").append(actionName));
+        }
+        mapOfMessages.put(messageIdForTrack, new StringBuilder().append(actionName));
+      }
+    }
+
+    if (!chainedToExisting && !TextUtils.isEmpty(fullEventName.toString())) {
+      trackEventWithMessageId(fullEventName.toString(), messageId, value, info, params);
+    } else if (TextUtils.isEmpty(fullEventName.toString())) {
```
Comment (milos1290) [inline AndroidSDK/src/com/leanplum/ActionContext.java]: if `fullEventName.toString()` is null or empty we won't track anything ?

Comment (Enuviel) [inline AndroidSDK/src/com/leanplum/ActionContext.java]: yep.

Comment (milos1290) [inline AndroidSDK/src/com/leanplum/ActionContext.java]: OK, can we check first for `TextUtils.isEmpty(fullEventName.toString()` and log the message before doing `if/else` ?

**Comment by Enuviel** [comment]
As discussed yesterday, we will have to each message only two events: view and Accept.

**Pull request merged**